### PR TITLE
[Merged by Bors] - Ensure doppelganger detects attestations in blocks

### PR DESCRIPTION
**Pull request by paulhauner**
## Issue Addressed

NA

## Proposed Changes

When testing our (not-yet-released) Doppelganger implementation, I noticed that we aren't detecting attestations included in blocks (only those on the gossip network).

This is because during [block processing](https://github.com/sigp/lighthouse/blob/e8c0d1f19b2736efb83c67a247e0022da5eaa7bb/beacon_node/beacon_chain/src/beacon_chain.rs#L2168) we only update the `observed_attestations` cache with each attestation, but not the `observed_attesters` cache. This is the correct behaviour when we consider the [p2p spec](https://github.com/ethereum/eth2.0-specs/blob/v1.0.1/specs/phase0/p2p-interface.md):

> [IGNORE] There has been no other valid attestation seen on an attestation subnet that has an identical attestation.data.target.epoch and participating validator index.

We're doing the right thing here and still allowing attestations on gossip that we've seen in a block. However, this doesn't work so nicely for Doppelganger.

To resolve this, I've taken the following steps:

- Add a `observed_block_attesters` cache.
- Rename `observed_attesters` to `observed_gossip_attesters`.

## TODO

- [x] Add a test to ensure a validator that's been seen in a block attestation (but not a gossip attestation) returns `true` for `BeaconChain::validator_seen_at_epoch`.
- [x] Add a test to ensure `observed_block_attesters` isn't polluted via gossip attestations and vice versa.

### CODE REVIEW

File: beacon_node/beacon_chain/src/beacon_chain.rs
Comment (realbigsean) [inline beacon_node/beacon_chain/src/beacon_chain.rs]: I think `OBSERVED_ATTESTERS_EPOCH_CAPACITY` would be a little clearer, to distinguish this from the total number of attesters in the cache

File: beacon_node/beacon_chain/src/beacon_chain.rs
Comment (realbigsean) [inline beacon_node/beacon_chain/src/beacon_chain.rs]:
```suggestion
                            "Failed to register observed block attester";
```

File: beacon_node/beacon_chain/src/beacon_chain.rs
Comment (realbigsean) [reply beacon_node/beacon_chain/src/beacon_chain.rs]: Also the cache's `lowest_permissible_epoch` (which it prunes based on)  is based off of the `target_epoch - (MAX_CAPACITY - 1)` so this check might have too wide a range.

**Review by realbigsean — COMMENTED** [review]
Generally wondering if we are OK with creating and populating the new cache even when doppelganger is disabled?

**Comment by paulhauner** [comment]
All comments addressed, thanks for the review @realbigsean.

I'll wait for your approval before merging this one :)

**Review by realbigsean — APPROVED** [review]
Looks good to me!

**Pull request closed without merging**
